File: 0207-course-schedule/0207-course-schedule.java
class Solution {
    HashMap<Integer, ArrayList<Integer>> map = new HashMap<>();
    HashSet<Integer> visited = new HashSet<>();
    HashSet<Integer> path = new HashSet<>();

    public boolean canFinish(int numCourses, int[][] prerequisites) {

        for (int i = 0; i < prerequisites.length; i++) {
            map.put(prerequisites[i][0], map.getOrDefault(prerequisites[i][0], new ArrayList<Integer>()));
            map.get(prerequisites[i][0]).add(prerequisites[i][1]);
        }

        for (var c : map.entrySet()) {
            if (!dfs(c.getKey())) return false;
        }

        return true;
    }

    private boolean dfs(int a) {
        if (path.contains(a)) return false;
        if (visited.contains(a)) return true;

        visited.add(a);
        path.add(a);

        if (!map.containsKey(a)) {
            path.remove(a);
            return true;
        }
        var curr = map.get(a);

        for (int num : curr) {
            if (!dfs(num)) return false;
        }

        path.remove(a);
        return true;
    }
}

File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        unordered_map<int, vector<int>> map;

        for (auto i{0uz}; i < prerequisites.size(); ++i) {
            map[prerequisites[i][0]].push_back(prerequisites[i][1]);
        }

        unordered_set<int> visited;
        unordered_set<int> path;

        for (auto [pre, _] : map) {
            if (!dfs(map, visited, path, pre)) return false;
        }

        return true;
    }

    bool dfs(unordered_map<int, vector<int>>& map, unordered_set<int>& visited, unordered_set<int>& path, int curr) {
        if (path.count(curr)) return false;
        if (visited.count(curr)) return true;
        path.insert(curr);
        visited.insert(curr);

        for (auto i{0uz}; i < map[curr].size(); ++i) {
            if (!dfs(map, visited, path, map[curr][i])) return false;
        }
        
        path.erase(curr);

        return true;
    }
};